### DISCUSSION TITLE
Update Anaconda packages to postgres 14/15 and latest RDKit release

### DISCUSSION THREAD

**Discussion by JamieAtkins** [rdkit/rdkit#6283] [Q&A] [answered]
Hello, I am currently attempting to install the RDKit cartridge into postgres using the Anaconda method described in the installation instructions on Windows. 
This has been successful, however the postgres version is 11 and the RDKit version is from 2019. I have been looking at trying to update the install package myself but unfortunately I am a complete novice in this area. 
I wondered if it would be possible to update the Windows and Linux packages available from the Anaconda site with up to date versions of each? 
I see that others have asked for this also and did not know whether I should add to their requests or create my own.
Thank you

**Comment by kienerj** [comment]
Hi Flingoo,

I agree it would be nice to have this.

But since it wasn't available I looked at building git myself (on windows) and it actually isn't that hard. see this discussion:

https://github.com/rdkit/rdkit/discussions/6148

It worked for me like this on windows 10.

The "isn't that hard" part however greatly depends on your existing PCs setup and what privileges you have on it. Most notably if you can easily install msvc (eg Visual studio) or not.

For Linux the recent blog post should probbaly get you started:

https://greglandrum.github.io/rdkit-blog/posts/2023-03-17-setting-up-a-cxx-dev-env2.html

To build the cartridge is just a matter of adding the according option to make: `-D RDK_BUILD_PGSQL=ON`

**Reply by JamieAtkins** [reply]
Hi kienrj,

Thank you for your response.

Were you able to create a tarball that could be used to deploy this on multiple Windows machines? I am having a great deal of difficulty with this. I've tried following the instructions in the blog post suggested above but I am stuck at the "make -j4 install" step. I installed make using chocolatey but the cmake command does not create a MakeFile. I tried building the project with Visual Studio but that appears to have got stuck. I'm sure these problems are all a result of my lack of familiarity with these technologies.

**Reply by kienerj** [reply]
> Were you able to create a tarball that could be used to deploy this on multiple Windows machines? 

No. This would be purely a "local install". 

I'm wondering why you would want to do that? For easy setup of a dev environment on developers local machine? I imagine you could "hack" something together (python or batch script) that create an appropriate conda environment, moves the recompiled rdkit files at the right places and sets the right PATH. But for sure would require some trial and error. but you only really need it for the cartridge to get exactly the combo of PostgreSQL and RDKit versions you need.

**Reply by JamieAtkins** [reply]
The reason I want to be able to do this is that the tarball currently available on the Anaconda website allows me to create an environment with a fully functional postgres database with the RDKit cartridge installed into it in about 4 commands and takes less than 3 minutes. This is incredibly convenient and can be done on any Windows PC in an almost entirely automated process. If the process to create that file already exists, I'd like to be able to do it but for more recent versions of Postgres/RDKit.

**Reply by kienerj** [reply]
I just added what I did below. Theoretically this could very easily be put in a batch script and hence also automated. At least after the prerequisites are installed.

**Comment by kienerj** [comment, marked as answer]
Here is the guide I wrote for myself to build RDKit on Windows. This is almost a copy&paste from #6148

# Building RDKit Cartridge on Windows

## Foreword

Currently there is no one clear guide available to perform this. This guide is based on these links:

- [[Guide for Linux](https://greglandrum.github.io/rdkit-blog/posts/2023-03-17-setting-up-a-cxx-dev-env2.html)](https://greglandrum.github.io/rdkit-blog/posts/2023-03-17-setting-up-a-cxx-dev-env2.html)
- [[Guide for Windows 11 - only rdkit](https://github.com/rdkit/rdkit/discussions/6148)](https://github.com/rdkit/rdkit/discussions/6148)
- [[Guide for Windows with installed PostgreSQL - requires admin privs](https://github.com/rdkit/rdkit/tree/master/Code/PgSQL/rdkit)](https://github.com/rdkit/rdkit/tree/master/Code/PgSQL/rdkit)

This guide is based on my laptop with Win 10 1909 installed.

In fact this guide as-is will build rdkit, all the wrappers an the cartridge. So the "complete package". A bit overkill.

## Preparation

Visual Studio 19 was already installed on my laptop. You will need to install a version of Visual Studio and get the build tools for your version of Visual Studio which include msvc and cmake. You also need to check which additional older versions of MSVC you need,. The version of visual studio doesn't really matter, what matters is the correct version of MSVC that is compatible with your build environment. Going by [this announcement](https://conda-forge.org/docs/user/announcements.html) from conda-forge, that currently should be msvc v142 (VS 2019). Therefore if you install VS 2022 you need to select msvc v142 under the optional installation parts of the builds tools installer. 

Hope that was clear enough.

### Clone git repository

Change to directory of your whish and clone rdkit. change path below.

```bash
cd C:\path\to\repository
git clone https://github.com/rdkit/rdkit.git rdkit
cd rdkit
mkdir build
```

### Create Anaconda environment

Guides all prefer to use mamba so we need to install that first:

```bash
conda install mamba -n base -c conda-forge
```

Then we create a new environment for building rdkit and rdkit cartridge

```bash
mamba create -n py310_rdkit_build -c conda-forge python=3.10 boost-cpp boost cairo pandas pillow freetype cmake numpy eigen matplotlib swig pkg-config postgresql
```

### libcairo

If you are interested in the python build of rdkit, you might need to add `libcairo-2.dll` to the PATH. You can get it via MYSYS2 or [[uniconverter](https://sk1project.net/uc2/download/)](https://sk1project.net/uc2/download/). In the later case it can be extracted from the msi and put in the path.

```bash
msiexec /a uniconvertor-2.0rc5-win64_headless.msi TARGETDIR=uniconvertor-2.0rc5-win64_headless /qn
```

You can then find `libcairo-2.dll` in `uniconvertor-2.0rc5-win64_headless\PFiles\UniConvertor-2.0rc5\dlls` and then paste it somewhere in the PATH for example `%CONDA_PREFIX%\Library\bin`.

## Building

In an anaconda prompt, activate the above environment and cd into the build dir:

```bash
mamba activate py310_rdkit_build
cd C:\path\to\repository\rdkit\build
```

Then configure the build:

```bash
cmake -DPy_ENABLE_SHARED=1 ^
-DBOOST_ROOT="%CONDA_PREFIX%" ^
-DBoost_NO_SYSTEM_PATHS=ON ^
-DBoost_NO_BOOST_CMAKE=TRUE ^
-DRDK_INSTALL_INTREE=ON ^
-DRDK_INSTALL_STATIC_LIBS=OFF ^
-DRDK_BUILD_CPP_TESTS=ON ^
-DRDK_BUILD_CAIRO_SUPPORT=ON ^
-DRDK_BUILD_FREETYPE_SUPPORT=ON ^
-DBOOST_ROOT=$CONDA_PREFIX ^
-DRDK_BOOST_PYTHON3_NAME=python310 ^
-DBoost_INCLUDE_DIR=%CONDA_PREFIX%/Library/include ^
-DRDK_BUILD_INCHI_SUPPORT=ON ^
-DRDK_BUILD_YAEHMOP_SUPPORT=ON ^
-DRDK_BUILD_XYZ2MOL_SUPPORT=ON ^
-D RDK_INSTALL_DLLS_MSVC=ON ^
-D RDK_BUILD_PGSQL=ON ^
-D RDK_PGSQL_STATIC=OFF ^
..

```

Finally run the build:

```bash
cmake --build . --config Release --target install -j 6
```

To use the new rdkit build from python, the dlls need to be explicitly imported. Adding the directories to PATH won't help:

```python
import os
os.add_dll_directory(r"C:\path\to\repository\rdkit\lib")
```

This is due to a change in Python 3.8:

> DLL dependencies for extension modules and DLLs loaded with [`[ctypes](https://docs.python.org/3/library/ctypes.html#module-ctypes)`](https://docs.python.org/3/library/ctypes.html#module-ctypes) on Windows are now resolved more securely. Only the system paths, the directory containing the DLL or PYD file, and directories added with [`[add_dll_directory()](https://docs.python.org/3/library/os.html#os.add_dll_directory)`](https://docs.python.org/3/library/os.html#os.add_dll_directory) are searched for load-time dependencies. **Specifically, `PATH` and the current working directory are no longer used, and modifications to these will no longer have any effect on normal DLL resolution.**

**Reply by JamieAtkins** [reply]
Thank you very much, I will see if I can recreate this locally and report back probably tomorrow now 👍

**Reply by JamieAtkins** [reply]
Good morning, thank you for your continued support on this. I have followed the process as described but am failing at the final step.
`created extension rdkit;`
I am seeing the error message described on the RDKit installation site stating
`ERROR: could not load library "<POSTGRES_LOCATION>/lib/rdkit.dll": The specified module could not be found.`
I have tried adding Boost lib locations to my PATH and running the add_dll_directory section in your instructions and it still doesn't work. At this point either I'm adding the wrong location for the libs, or there's something wrong with my rdkit.dll file (which is there). The generated file is only 183kb, but the one from an older installation is over 3mb. I was wondering if that might be the problem.

**Reply by JamieAtkins** [reply]
@kienerj I have it working, I needed to manually copy the dll files from the build to the environment (as you suggested in https://github.com/rdkit/rdkit/discussions/6148) I will repeat and make sure to document anywhere I deviated from your instructions, but I think it was just that. Once again, thank you for your help in this.

**Reply by kienerj** [reply]
Yeah with all the playing around some steps might get lost. With Trial & error:
To get rid of the error you must either move the generated dlls to the path or add `build\bin\Release` to `PATH`.

``` bash
set RDBASE=C:\path\to\rdkit
set PATH=%PATH%;%RDBASE%\build\bin\Release
pg_ctl -D rdkit -l logfile start
```

**Comment by greglandrum** [comment]
I agree that it would be good to have conda builds of rdkit-postgresql. I would not do this via the rdkit channel (which requires me to manually do builds) but to use the conda-forge infrastructure instead.

@peterbygrave has started to look into this already: https://github.com/conda-forge/rdkit-feedstock/issues/128

**Reply by JamieAtkins** [reply]
Thanks @greglandrum with the help of @kienerj I have been able to create my own conda environments fairly easily now, but the ability for others that have not been through this process to be able to simply install an up to date package would be fantastic. I will keep my eyes open for further developments.
